<center>
<h1> Assignment 3: Regression Analysis</h1>
<hr>
<h2>Predict The Bay Area’s Home Prices</h2>
<hr>
<h3> MCIS 6283-Machine Learning </h3>

<h3><mark>Due date: Mar 9th, 2021 (Tuesday)</mark></h3>
<h3>Total Points: 100</h3>

<h4>Instructor: Dr. Ahmad Al Shami</h4>


</center>

### Please put your name, student ID, date and time here (5 points)
* Name:Sowmya Reddy Baddam
* Student ID:999900971
* Date:03/05/2021
* Time:5.00 p.m

### Instructions
* In this assignment, you will investigate the housing price dataset from Bay Area Home Sales Database and Zillow to Predict The Bay Area’s Home Prices.
* This dataset was based on the homes sold between January 2013 and December 2015.
* Please use all the techniques we have learned in the class to preprocesss/clean the dataset.
* You can drop redundant features.
<p style="color:blue"><b>final_data.csv</b></p>
* Please apply the folowing four methods to Predict The Bay Area’s Home Prices.

<h3>Method 1: Multiple Linear Regression</h3>

* Bulding the optimal model using Automatic implementations of Backward Elimination
```
def backwardElimination(x, sl):
    numVars = len(x[0])
    for i in range(0, numVars):
        regressor_OLS = sm.OLS(y, x).fit()
        maxVar = max(regressor_OLS.pvalues)
        if maxVar > sl:
            for j in range(0, numVars - i):
                if (regressor_OLS.pvalues[j] == maxVar):
                    x = np.delete(x, j, 1)
    regressor_OLS.summary()
    return x
```
* <mark>Split the dataset into training sets and test sets</mark>

* Fit LinearRegression to the training sets 
* Print the Linear Regression R squared score on the test sets
* Print the Linear Regression RMSE score on the test sets

<h3>Method 2: Least Absolute Shrinkage and Selection Operator(Lasso)</h3>

* Fit Lasso to the training sets. 
* Print the Lasso R squared score on the test sets
* Print the Lasso RMSE score on the test sets

<h3>Method 3: Decision Tree </h3>

* Fit Decision Tree to the training set. 
* Print the Decision Tree R squared score on the test sets
* Print the Decision Tree RMSE score on the test sets

<h3>Method 4: Random Forest </h3>

* Fit Random Forest to the training sets. 
* Print the Random Forest R squared score on the test sets
* Print the Random Forest RMSE score on the test sets

## 1. Import dataset "final_data.csv" in pandas (5 points)

In [145]:
import pandas as pd
dataset = pd.read_csv('final_data.csv')
print(dataset.columns)
print(dataset.shape)



Index(['Unnamed: 0', 'address', 'info', 'z_address', 'bathrooms', 'bedrooms',
       'finishedsqft', 'lastsolddate', 'lastsoldprice', 'latitude',
       'longitude', 'neighborhood', 'totalrooms', 'usecode', 'yearbuilt',
       'zestimate', 'zindexvalue', 'zipcode', 'zpid'],
      dtype='object')
(11330, 19)


In [146]:
print(dataset.info())
#print(dataset.dtypes)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11330 entries, 0 to 11329
Data columns (total 19 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Unnamed: 0     11330 non-null  int64  
 1   address        11330 non-null  object 
 2   info           11330 non-null  object 
 3   z_address      11330 non-null  object 
 4   bathrooms      11330 non-null  float64
 5   bedrooms       11330 non-null  float64
 6   finishedsqft   11330 non-null  float64
 7   lastsolddate   11330 non-null  object 
 8   lastsoldprice  11330 non-null  float64
 9   latitude       11330 non-null  float64
 10  longitude      11330 non-null  float64
 11  neighborhood   11330 non-null  object 
 12  totalrooms     11330 non-null  float64
 13  usecode        11330 non-null  object 
 14  yearbuilt      11330 non-null  float64
 15  zestimate      11330 non-null  float64
 16  zindexvalue    11330 non-null  object 
 17  zipcode        11330 non-null  float64
 18  zpid  

In [147]:
dataset.head()

Unnamed: 0                             address  \
0           2  Address: 1160 Mission Street #2007   
1           5       Address: 260 King Street #475   
2           7     Address: 560 Missouri Street #B   
3           9        Address: 350 Missouri Street   
4          11         Address: 3658 Folsom Street   

                                                info  \
0   San FranciscoSales price: 1300000Sales date: ...   
1   San FranciscoSales price: 750000Sales date: 0...   
2   San FranciscoSales price: 1495000Sales date: ...   
3   San FranciscoSales price: 2700000Sales date: ...   
4   San FranciscoSales price: 1530000Sales date: ...   

                   z_address  bathrooms  bedrooms  finishedsqft lastsolddate  \
0  1160 Mission St UNIT 2007        2.0       2.0        1043.0   02/17/2016   
1       260 King St UNIT 475        1.0       1.0         903.0   02/17/2016   
2        560 Missouri St # B        4.0       3.0        1425.0   02/17/2016   
3            350 Missouri St        3.0       3.0        2231.0   02/17/2016   
4             3658 Folsom St        3.0       3.0        1300.0   02/17/2016   

   lastsoldprice   latitude   longitude     neighborhood  totalrooms  \
0      1300000.0  37.778705 -122.412635  South of Market         4.0   
1       750000.0  37.777641 -122.393417  South of Market         3.0   
2      1495000.0  37.759198 -122.396516     Potrero Hill         6.0   
3      2700000.0  37.761886 -122.396769     Potrero Hill        10.0   
4      1530000.0  37.740795 -122.413453   Bernal Heights         4.0   

        usecode  yearbuilt  zestimate zindexvalue  zipcode        zpid  
0   Condominium     2007.0  1167508.0     975,700  94103.0  83152781.0  
1   Condominium     2004.0   823719.0     975,700  94107.0  69819817.0  
2   Condominium     2003.0  1708594.0   1,277,600  94107.0  64972847.0  
3  SingleFamily     1927.0  2411236.0   1,277,600  94107.0  15149489.0  
4  SingleFamily     1900.0  1918539.0   1,248,000  94110.0  15161978.0

## 2. Preprocess/Clean the Dataset(20 points)

In [148]:
#converting datatypes of lastsolddate column to date and zindexvalue to int
dataset['lastsolddate']=pd.to_datetime(dataset['lastsolddate'])
dataset['zindexvalue']=dataset['zindexvalue'].str.replace(',','')



In [149]:
dataset['zindexvalue']=pd.to_numeric(dataset['zindexvalue'])
print(dataset.dtypes)


Unnamed: 0                int64
address                  object
info                     object
z_address                object
bathrooms               float64
bedrooms                float64
finishedsqft            float64
lastsolddate     datetime64[ns]
lastsoldprice           float64
latitude                float64
longitude               float64
neighborhood             object
totalrooms              float64
usecode                  object
yearbuilt               float64
zestimate               float64
zindexvalue               int64
zipcode                 float64
zpid                    float64
dtype: object


In [150]:
#print(dataset['lastsolddate'],dataset['zindexvalue'])

In [151]:
#* This dataset was based on the homes sold between January 2013 and December 2015. 
#This dataset has other date ranges also . So in order to make date ranges in the given range I have done this
start_date = '2013-01-02'
end_date = '2015-12-31'
mask = (dataset['lastsolddate'] > start_date) & (dataset['lastsolddate'] <= end_date)
dataset_new = dataset.loc[mask]
print(dataset_new.shape)

(9378, 19)


In [152]:
dataset_new.head()

Unnamed: 0                              address  \
12           24        Address: 370 Church Street #F   
66          136  Address: 861 Rhode Island Street #A   
94          253     Address: 701 Pennsylvania Avenue   
222         564         Address: 180 7th Street #408   
223         569        Address: 961 San Bruno Avenue   

                                                  info  \
12    San FranciscoSales price: 888000Sales date: 0...   
66    San FranciscoSales price: 1110000Sales date: ...   
94    San FranciscoSales price: 850000Sales date: 0...   
222   San FranciscoSales price: 550000Sales date: 1...   
223   San FranciscoSales price: 980000Sales date: 1...   

                      z_address  bathrooms  bedrooms  finishedsqft  \
12          370 Church St APT F        1.0       1.0         780.0   
66         961A Rhode Island St        2.5       4.0        2166.0   
94   701 Pennsylvania Ave # T13        2.0       1.0         842.0   
222            180 7th St # 408        1.0       1.0         470.0   
223           961 San Bruno Ave        2.0       2.0        1297.0   

    lastsolddate  lastsoldprice   latitude   longitude  \
12    2015-08-12       863000.0  37.765062 -122.429077   
66    2015-06-11      1600000.0  37.757681 -122.401578   
94    2015-03-24       640000.0  37.757380 -122.392830   
222   2015-12-31       550000.0  37.778082 -122.410082   
223   2015-12-31       980000.0  37.757433 -122.404508   

                                 neighborhood  totalrooms          usecode  \
12   Eureka Valley - Dolores Heights - Castro         4.0      Condominium   
66                               Potrero Hill        11.0  MultiFamily2To4   
94                               Potrero Hill         2.0      Condominium   
222                           South of Market         3.0      Condominium   
223                                   Mission         6.0     SingleFamily   

     yearbuilt  zestimate  zindexvalue  zipcode        zpid  
12      1981.0   847894.0      1744100  94114.0  15143549.0  
66      1908.0  1599122.0      1277600  94107.0  15150449.0  
94      1999.0   905401.0      1277600  94107.0  58770199.0  
222     1992.0   540858.0       975700  94103.0  15147378.0  
223     1927.0  1100483.0      1185400  94110.0  15151012.0

In [153]:
# since use code has 10 variables that needs to be encoded we have done enocdding using one-hot encoding
#print(dataset['usecode'].value_counts())
#print(dataset['neighborhood'].value_counts())

In [154]:
#Normalization
#since all the features have variance in the same range we are not doing any normalization
print(dataset.var())
#Normalization
#import numpy as np
#print(dataset.var())
#dataset['finishedsqft'] = np.log(dataset['finishedsqft'])
#dataset['bedrooms'] = np.log(dataset['bedrooms'])
#dataset['bathrooms']=np.log(dataset['bathrooms'])
#dataset['totalrooms'] = np.log(dataset['totalrooms'])
#dataset['zindexvalue'] = np.log(dataset['zindexvalue'])


Unnamed: 0       2.422550e+07
bathrooms        1.096959e+00
bedrooms         1.688589e+00
finishedsqft     8.500439e+05
lastsoldprice    1.085929e+12
latitude         6.542351e-04
longitude        9.451565e-04
totalrooms       1.470355e+02
yearbuilt        1.437259e+03
zestimate        1.511466e+12
zindexvalue      3.420110e+11
zipcode          8.837649e+01
zpid             6.085157e+15
dtype: float64


In [155]:
Features=dataset[['finishedsqft','totalrooms','zindexvalue']]
#Features=dataset.drop(['lastsoldprice','Unnamed: 0','address','info','z_address','lastsolddate','latitude','longitude','neighborhood','usecode','yearbuilt','totalrooms','zipcode','zpid','bathrooms','bedrooms'],axis=1)
Labels=dataset['lastsoldprice']
print(Features.columns)
print(Features.head())

Index(['finishedsqft', 'totalrooms', 'zindexvalue'], dtype='object')
   finishedsqft  totalrooms  zindexvalue
0        1043.0         4.0       975700
1         903.0         3.0       975700
2        1425.0         6.0      1277600
3        2231.0        10.0      1277600
4        1300.0         4.0      1248000


In [156]:
#Since these features have high standard deviation we are standrazing them
from sklearn.preprocessing import StandardScaler
import pandas as pd
scaler = StandardScaler()
print(dataset.describe())
Features=dataset[['finishedsqft','totalrooms','zindexvalue']]
Features= scaler.fit_transform(Features);
print(pd.DataFrame(Features).describe())


         Unnamed: 0     bathrooms      bedrooms  finishedsqft  lastsoldprice  \
count  11330.000000  11330.000000  11330.000000  11330.000000   1.133000e+04   
mean    9171.729214      1.980229      2.614475   1585.420918   1.263928e+06   
std     4921.941074      1.047358      1.299457    921.978245   1.042079e+06   
min        2.000000      0.500000      0.000000      1.000000   5.350000e+02   
25%     5039.750000      1.000000      2.000000   1019.000000   7.292500e+05   
50%     9198.500000      2.000000      2.000000   1362.000000   9.900000e+05   
75%    13374.750000      2.000000      3.000000   1876.000000   1.450000e+06   
max    17632.000000     14.000000     20.000000  27275.000000   2.388900e+07   

           latitude     longitude    totalrooms     yearbuilt     zestimate  \
count  11330.000000  11330.000000  11330.000000  11330.000000  1.133000e+04   
mean      37.759711   -122.436518      6.111562   1948.498147  1.565695e+06   
std        0.025578      0.030743     12.1

In [157]:
from sklearn.feature_extraction.text import TfidfVectorizer
vec=TfidfVectorizer()
text_tfidf=vec.fit_transform(dataset['address'])
print(type(text_tfidf))
print(text_tfidf.toarray().shape)
print(pd.DataFrame(text_tfidf.toarray()).columns)
#print(pd.DataFrame(text_tfidf.toarray()).head())
#print(vec.vocabulary_)

<class 'scipy.sparse.csr.csr_matrix'>
(11330, 4349)
RangeIndex(start=0, stop=4349, step=1)


In [158]:
# In order to find the relation between last sold price and all other columns in dataset we have find corr
corr=dataset.corr()
corr['lastsoldprice'].sort_values(ascending=False)

lastsoldprice    1.000000
zestimate        0.876939
finishedsqft     0.647208
bathrooms        0.536880
zindexvalue      0.460429
bedrooms         0.395478
latitude         0.283107
totalrooms       0.093527
zipcode          0.018839
zpid            -0.038444
longitude       -0.052595
Unnamed: 0      -0.134305
yearbuilt       -0.189055
Name: lastsoldprice, dtype: float64

In [159]:
# since use code has 10 variables that needs to be encoded we have done enocdding using one-hot encoding
# we don't want our model ti get influenced by zillow estimate , so we are dropinng zestimate
import pandas as pd
X=dataset.drop(['lastsoldprice','Unnamed: 0','address','info','zestimate','z_address','lastsolddate','latitude','longitude','neighborhood','yearbuilt','zipcode','zpid'],axis=1)
X_new=pd.get_dummies(X,['usecode'])
y=dataset['lastsoldprice']

In [160]:
''' Since these features have high correlation and are suitable for the model to predict the price of new houses, I have dropped all other columns from datset before training a model''' 
X_new.head()

bathrooms  bedrooms  finishedsqft  totalrooms  zindexvalue  \
0        2.0       2.0        1043.0         4.0       975700   
1        1.0       1.0         903.0         3.0       975700   
2        4.0       3.0        1425.0         6.0      1277600   
3        3.0       3.0        2231.0        10.0      1277600   
4        3.0       3.0        1300.0         4.0      1248000   

   usecode_Apartment  usecode_Condominium  usecode_Cooperative  \
0                  0                    1                    0   
1                  0                    1                    0   
2                  0                    1                    0   
3                  0                    0                    0   
4                  0                    0                    0   

   usecode_Duplex  usecode_Miscellaneous  usecode_Mobile  \
0               0                      0               0   
1               0                      0               0   
2               0                      0               0   
3               0                      0               0   
4               0                      0               0   

   usecode_MultiFamily2To4  usecode_MultiFamily5Plus  usecode_SingleFamily  \
0                        0                         0                     0   
1                        0                         0                     0   
2                        0                         0                     0   
3                        0                         0                     1   
4                        0                         0                     1   

   usecode_Townhouse  
0                  0  
1                  0  
2                  0  
3                  0  
4                  0

## 3. Method 1: Multiple Linear Regression (15 points)

In [161]:
import statsmodels.formula.api as sm
def backwardElimination(x, sl):
    
    numVars = len(x[0])
    for i in range(0, numVars):
        regressor_OLS = sm.OLS(y, x).fit()
        maxVar = max(regressor_OLS.pvalues)
        if maxVar > sl:
            for j in range(0, numVars - i):
                if (regressor_OLS.pvalues[j] == maxVar):
                     x = np.delete(x, j, 1)
    regressor_OLS.summary()
    return x

In [162]:
import statsmodels.api as sm
import numpy as np
SL=0.05
#X_new=X.drop(['lastsoldprice'],axis=1)
#print(X_new.shape)
X = np.append(arr = np.ones((11330,1)), values = X_new, axis = 1)
print(X.shape)

X_Modeled = backwardElimination(X, SL)
print(X_Modeled.shape)
print(X_Modeled)


(11330, 16)
(11330, 9)
[[1. 2. 2. ... 0. 0. 0.]
 [1. 1. 1. ... 0. 0. 0.]
 [1. 4. 3. ... 0. 0. 0.]
 ...
 [1. 1. 2. ... 0. 0. 1.]
 [1. 1. 2. ... 0. 0. 1.]
 [1. 2. 3. ... 0. 0. 1.]]


In [163]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_Modeled, 
                                                    y, test_size = 0.3, random_state = 0)

In [164]:
from sklearn.linear_model import LinearRegression
import numpy as np
reg=LinearRegression()

reg.fit(X_train,y_train)
print('Multiple Linear Regression R squared: %.4f'%reg.score(X_test, y_test))
y_pred=reg.predict(X_test)
#print(y_pred)
from sklearn.metrics import mean_squared_error
mse=mean_squared_error(y_pred,y_test)
rmse=np.sqrt(mse)
print('Multiple Linear Regression RMSE: %.4f'%rmse)

Multiple Linear Regression R squared: 0.5304
Multiple Linear Regression RMSE: 637826.2514


In [165]:
print(reg.coef_)

[ 0.00000000e+00  2.35571093e+05 -1.17640561e+05  5.94403247e+02
  4.51122127e-01  1.18319996e+05 -7.49982274e+05 -2.44140085e+05
  1.46987485e+05]


In [166]:
print(dict(zip(X_new.columns,reg.coef_)))

{'bathrooms': 0.0, 'bedrooms': 235571.09297913968, 'finishedsqft': -117640.56057189814, 'totalrooms': 594.403247346476, 'zindexvalue': 0.45112212710997424, 'usecode_Apartment': 118319.99554625533, 'usecode_Condominium': -749982.2735419082, 'usecode_Cooperative': -244140.08485362338, 'usecode_Duplex': 146987.4849078084}


## 4. Method 2: Least Absolute Shrinkage and Selection Operator(Lasso) (5 points bonus)

In [167]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X_new,y,test_size=0.3,random_state=0)

In [168]:
from sklearn.linear_model import Lasso
lasso=Lasso(alpha=0.1,max_iter=1e6)
lasso.fit(X_train,y_train)
print('Lasso R squared: %.4f'%lasso.score(X_test, y_test))

Lasso R squared: 0.5294


In [169]:
y_pred=lasso.predict(X_test)
#print(y_pred)
from sklearn.metrics import mean_squared_error
mse=mean_squared_error(y_pred,y_test)
rmse=np.sqrt(mse)
print('Lasso RMSE: %.4f'%rmse)

Lasso RMSE: 638524.6192


In [170]:
print(lasso.coef_)

[ 2.35399916e+05 -1.18643836e+05  5.92739621e+02  1.43616453e+03
  4.50956706e-01  0.00000000e+00  6.61073822e+04 -4.02687399e+05
 -8.06285505e+05  5.51267698e+04  4.53783549e+05 -2.98805403e+05
 -2.55769843e+05  9.41402632e+04 -8.74593569e+04]


In [171]:
print(dict(zip(X_new.columns,lasso.coef_)))

{'bathrooms': 235399.9160681771, 'bedrooms': -118643.83601647888, 'finishedsqft': 592.7396212827373, 'totalrooms': 1436.1645301972285, 'zindexvalue': 0.4509567056284082, 'usecode_Apartment': 0.0, 'usecode_Condominium': 66107.38222160892, 'usecode_Cooperative': -402687.3993893953, 'usecode_Duplex': -806285.505219038, 'usecode_Miscellaneous': 55126.76976656638, 'usecode_Mobile': 453783.54855204176, 'usecode_MultiFamily2To4': -298805.4025557608, 'usecode_MultiFamily5Plus': -255769.84340550101, 'usecode_SingleFamily': 94140.26317772268, 'usecode_Townhouse': -87459.35687638387}


## 5. Method 3: Decision Tree (20 points)

In [172]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X_new,y,test_size=0.3,random_state=0)

In [173]:
from sklearn.tree import DecisionTreeRegressor
reg=DecisionTreeRegressor(random_state=0)
reg.fit(X_train,y_train)
print('Decision Tree Regression R squared: %.4f' % reg.score(X_test,y_test))



Decision Tree Regression R squared: 0.3334


In [174]:
from sklearn.metrics import mean_squared_error
import numpy as np
y_pred=reg.predict(X_test)
print(y_pred)
mse=mean_squared_error(y_pred,y_test)
rmse=np.sqrt(mse)
print('Decision Tree Regression RMSE: %.4f'%rmse)

[1580000. 1625000.  680000. ... 2750000. 3825000.  825000.]
Decision Tree Regression RMSE: 759937.3233


In [175]:
important_features=reg.feature_importances_
print(important_features)

[7.28664951e-02 3.51858286e-02 5.83136165e-01 5.96319026e-02
 2.10852902e-01 4.24634532e-05 4.63816805e-03 0.00000000e+00
 8.01543453e-04 2.21999857e-04 1.76158639e-05 6.41696561e-04
 0.00000000e+00 3.17205452e-02 2.42674734e-04]


In [176]:
print(dict(zip(X_new.columns,important_features)))

{'bathrooms': 0.07286649507687067, 'bedrooms': 0.035185828567763554, 'finishedsqft': 0.5831361648225822, 'totalrooms': 0.05963190255266725, 'zindexvalue': 0.21085290182223748, 'usecode_Apartment': 4.2463453237975224e-05, 'usecode_Condominium': 0.004638168045278753, 'usecode_Cooperative': 0.0, 'usecode_Duplex': 0.0008015434529566541, 'usecode_Miscellaneous': 0.00022199985740104192, 'usecode_Mobile': 1.761586388361363e-05, 'usecode_MultiFamily2To4': 0.000641696561250549, 'usecode_MultiFamily5Plus': 0.0, 'usecode_SingleFamily': 0.03172054518940716, 'usecode_Townhouse': 0.00024267473446281502}


## 6. Method 4: Random Forest (20 points)

In [177]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X_new,y,test_size=0.3,random_state=0)

In [178]:
from sklearn.ensemble import RandomForestRegressor
reg=RandomForestRegressor(n_estimators = 1000,random_state = 0)
reg.fit(X_train,y_train)
print('Random Forest Regression R squared: %.4f' % reg.score(X_test,y_test))

Random Forest Regression R squared: 0.6559


In [179]:
from sklearn.metrics import mean_squared_error
import numpy as np
y_pred=reg.predict(X_test)
print(y_pred)
mse=mean_squared_error(y_pred,y_test)
rmse=np.sqrt(mse)
print('Random Forest Regression RMSE: %.4f'%rmse)

[1813043.         1481390.          749379.28896104 ... 1835478.3
 3545264.5         742260.1       ]
Random Forest Regression RMSE: 545972.1760


In [180]:
important_features=reg.feature_importances_
print(important_features)

[7.24769262e-02 3.25760598e-02 5.76176421e-01 6.41157706e-02
 2.12797613e-01 2.45598706e-05 7.02090678e-03 1.41943987e-05
 1.45677930e-03 2.28236658e-04 2.40614514e-05 2.46192994e-03
 6.31897329e-06 3.04453433e-02 1.74878745e-04]


In [181]:
print(dict(zip(X_new.columns,important_features)))

{'bathrooms': 0.07247692624244702, 'bedrooms': 0.03257605981706971, 'finishedsqft': 0.5761764211230682, 'totalrooms': 0.06411577062959323, 'zindexvalue': 0.21279761272349626, 'usecode_Apartment': 2.4559870595640036e-05, 'usecode_Condominium': 0.007020906777563412, 'usecode_Cooperative': 1.4194398674119848e-05, 'usecode_Duplex': 0.0014567793030302737, 'usecode_Miscellaneous': 0.00022823665817676368, 'usecode_Mobile': 2.406145141550346e-05, 'usecode_MultiFamily2To4': 0.002461929940797956, 'usecode_MultiFamily5Plus': 6.31897329200874e-06, 'usecode_SingleFamily': 0.030445343346036293, 'usecode_Townhouse': 0.0001748787447435097}


## 7. Test Data Set Prediction Results Comparison Table of Method 1, Method 2, Method 3 and Method 4 (15 points)
* Please fill in your results to the following table

<table>
<thead>
    <tr>
    <th colspan="4">Predict The Bay Area’s Home Prices</th>
    </tr>
    <tr>
    <th> </th>
    <th>Multiple Linear Regression</th>
    <th>Lasso</th>
    <th>Decision Tree</th>
    <th>Random Forest</th>
    </tr>
</thead>
<tbody>
    <tr>
        <th>R Squared Score</th>
        <td> 0.5304 </td>
        <td> 0.5294 </td>
        <td> 0.3334</td>
        <td> 0.6559 </td>
    </tr>
     <tr>
         <th>RMSE</th>
        <td> 637826.2514 </td>
        <td> 638524.6192 </td>
        <td> 759937.3233 </td>
        <td> 545972.1760 </td>
    </tr>
</tbody>
</table>